include/coap2/net.h

```diff
@@ -758,7 +758,7 @@ coap_join_mcast_group(coap_context_t *ctx, const char *groupname);
  * @return Number of milliseconds spent in function or @c -1 if there was
  *         an error
  */
-int coap_io_process(coap_context_t *ctx, unsigned int timeout_ms);
+int coap_io_process(coap_context_t *ctx, int timeout_ms);
```

Comment:
Changing the data type of `timeout_ms` from unsigned to signed cuts down the maximum timeout from 49 days to 24 days. I am not sure we want to do this. Instead, I suggest keeping it unsigned and just reserve the all-one-bits value (i.e., -1) with special semantics. This would require some careful coding of comparisons, though.

Reply:
#462 pretty much covers this aspect. To overcome the weird "you cannot have a timeout of exactly 1 ms", we could do use -1 there as well ("the largest timeout value means "forever".)

Reply:
Currently, with timeout_ms == 0, this means indefinite wait until there is activity on a fd.  With timeout_ms == 1, this means (apart from the 1ms spin) that all outstanding fd activity is to be processed and then return without waiting.

We can change this to timeout_ms == 0 - never wait on fd (do not block)
timeout_ms == 0xffffffff (-1) - wait forever for fd activity (block)

or

We can leave timeout_ms == 0 - wait forever for fd activity (block)
timeout_ms == 0xffffffff (-1) - never wait on fd (do not block)

Going with the former means that we also need to special case coap_prepare_io*() functions which currently return 0 is there is nothing to do.

My preference would be for the later as a simpler change.

Reply:
Right. I prefer the simpler change as well.

Reply:
Is it worth changing
COAP_RUN_BLOCK to COAP_IO_WAIT
and
COAP_RUN_NOBLOCK to COAP_IO_NOWAIT
at this time?

Reply:
Seems reasonable.

Reply:
Grr - as it happens, epoll_wait() timeout parameter is an integer so we are limited to a maximum wait of 24 days if COAP_EPOLL_SUPPORT is defined - or indefinite until the next packet comes in.

Reply:
Code updated and pushed.